### DISCUSSION TITLE
JSON files included with package making it hard to bundle

### DISCUSSION THREAD

**Discussion by Goolyio** [apify/fingerprint-suite#276] [Ideas]
Hey, I've been trying to use this package within a lambda that is bundled using esbuild.

Unfortunatley the way the [JSON files](https://github.com/apify/fingerprint-suite/blob/master/packages/header-generator/src/header-generator.ts#L198) are included means that they are often not included in final builds when bundled and minified.

```ts
this.headersOrder = JSON.parse(readFileSync(`${__dirname}/data_files/headers-order.json`).toString());
```

Is there a reason these are JSON files? Could they not just be constant values within the javascript and you can optionally override them.

![image](https://github.com/apify/fingerprint-suite/assets/56264922/0d298a32-5509-4342-a08a-3fd0a24f34e6)

Currently I will need to inject these files manually into build outputs but try to make sure they are only included in bundles that use your package which is pretty janky.

**Comment by barjin** [comment]
Hello and thank you for your interest in this package!

For the time being, I converted this issue into a discussion, as I don't have a clear idea about the fix for this right now (and it's also not breaking the library... too much). 

This sounds quite similar to an issue we encountered quite a while ago (see #135). Back then, it was caused by the dynamic paths to the JSON files - the Vercel bundler couldn't follow those and therefore dropped those files. I'm assuming something similar is happening here - `esbuild`'s tree-shaking might skip `readFileSync` calls (not consider them actual references to the files?). I'm thinking that converting those to `require / import` calls might fix this, but that comes with its own issues we would need to solve before that.

As to your question about why are those files JSON - well, they are data files, and they are automatically generated. While it's possible to generate `.js` files (that are exporting the data), it seems to me like a very bad code smell. 

In the meantime, would you mind sharing details about your `esbuild` setup? I have only used it before as a part of larger, preconfigured projects, so I never really got to learn much about it. Do you have a minimal reproducible example?

**Comment by aecorredor** [comment]
+1 on this. I literally ran into this just today while deploying a scraper in AWS Lambda.

**Comment by aecorredor** [comment]
For the time being, I'm using `patch-package`to just inline the headers as JavaScript. Just curious, why do you think it'd be a code smell to generate the data files JavaScript files that just export a constant object? It could even use something like `Object.freeze` to make it immutable.

**Comment by aecorredor** [comment]
I actually just realized that the code is also depending on other more complicated data, like zip files, so nvm on the patching.

**Comment by aecorredor** [comment]
I'm actually going to try to include the files during the build step.

**Comment by aecorredor** [comment]
Yeap, that did it. You can just bundle the data files in your lambda code. I'm using pulumi to deploy, and here's an example of how you'd do it:

```ts
// Lives in deploy/handler.ts
import { createHandlerAssetArchive } from './utils.ts';

export const handler = new aws.lambda.Function('functionName', {
  role: role.arn,
  runtime: 'nodejs18.x',
  memorySize: 1024,
  code: createHandlerAssetArchive('../dist/handler.js', {
    extraContent: {
      data_files: new pulumi.asset.FileArchive(
        '../node_modules/header-generator/data_files'
      ),
    },
  }),
  timeout: 900,
  handler: 'index.handler',
});

// Helper in another file that I always use to deploy lambdas with pulumi
// Lives in deploy/utils.ts
import * as pulumi from '@pulumi/pulumi';

export const createHandlerAssetArchive = (
  handlerPath: string,
  opts?: {
    extraContent?: Record<
      string,
      pulumi.asset.FileAsset | pulumi.asset.FileArchive
    >;
  }
): pulumi.asset.AssetArchive => {
  return new pulumi.asset.AssetArchive({
    'index.js': new pulumi.asset.FileAsset(handlerPath),
    ...opts?.extraContent,
  });
};
```

**Comment by MattGould1** [comment]
Hi for anyone encountering this but using Serverless Framework (or CloudFormation) the way I solved this was by creating a Lambda Layer. I added the entire `fingerprint-injector` module as a layer rather than just the `header-generator` so modify to your requirements.

Create a directory for the layer, mine is in `layers/fingerprintInjector`. In that directory create a `package.json` with:

```json
{
  "name": "fingerprint-injector",
  "description": "fingerprint injector",
  "dependencies": {
    "fingerprint-injector": "2.1.49"
  }
}
```

Run `npm install` in the directory with your newly created `package.json`.

Add your layer to your `serverless.yml` file.

```yml
# The rest of your Serverless.yml file
layers:
  fingerprintInjector:
    path: 'layers/fingerprintInjector/'
  
functions:
  yourLambdaName:
    handler: ../index.handler
    layers:
      - !Ref FingerprintInjectorLambdaLayer
```

Finally mark `fingerprint-injector` as an `external` package in esbuild/webpack/other so that it's not included in your lambda. For me, I'm using `serverless-esbuild` so that looks like:

```js
{
      external: [
        'fingerprint-injector'
      ]
 }
 ```

**Reply by ollebergkvist** [reply]
How could one solve this with Next.js?

**Reply by maxbeatty** [reply]
this solved it for us https://nextjs.org/docs/app/api-reference/next-config-js/serverComponentsExternalPackages

```js
// See https://nextjs.org/docs/app/api-reference/next-config-js.
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    serverComponentsExternalPackages: [
      "fingerprint-generator",
    ],
  },
};
```

**Comment by OrigamingWasTaken** [comment]
This is still a very problematic issue when bundling code from a machine, and sending a compiled executable to users.
The bundled executable references to the json file path on my machine, so the file won't exist on other machines.
